app-ng/src/persistent-qvm.lisp

```diff
@@ -0,0 +1,147 @@
+;;;; api/persistent-qvm.lisp
+;;;;
+;;;; Author: appleby
+
+(in-package :qvm-app-ng)
+
+(deftype persistent-qvm-token () 'string)
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-empty-persistent-qvms-db ()
+    (make-hash-table :test 'equal)))
+
+(global-vars:define-global-var **persistent-qvms** (make-empty-persistent-qvms-db)
+  "The database of persistent QVMs. The keys are integers and the values are lists of (QVM LOCK METADATA) triples.")
+
+(global-vars:define-global-var **persistent-qvms-lock** (bt:make-lock "Persistent QVMs DB Lock"))
+
+(defun reset-persistent-qvms-db ()
+  (bt:with-lock-held (**persistent-qvms-lock**)
+    (setf **persistent-qvms** (make-empty-persistent-qvms-db))))
+
+(defun persistent-qvms-count ()
+  (bt:with-lock-held (**persistent-qvms-lock**)
+    (hash-table-count **persistent-qvms**)))
+
+(defmacro with-persistent-qvm ((qvm &optional metadata) token &body body)
+  (check-type qvm symbol)
+  (check-type metadata (or null symbol))
+  (when (null metadata)
+    (setf metadata (gensym "metadata")))
+  (alexandria:with-gensyms (lock)
+    (alexandria:once-only (token)
+      `(destructuring-bind (,qvm ,lock ,metadata) (%lookup-persistent-qvm-or-lose ,token)
+         (declare (ignorable ,qvm))
+         (bt:with-lock-held (,lock)
+           (cond ((%marked-for-deletion-p ,metadata)
+                  (error "Persistent QVM ~A is marked for deletion." ,token))
+                 (t ,@body)))))))
+
+(defun %insert-persistent-qvm-locked (token persistent-qvm)
+  (setf (gethash token **persistent-qvms**) persistent-qvm))
+
+(defun %remove-persistent-qvm-locked (token)
+  (remhash token **persistent-qvms**))
+
+(defun %remove-persistent-qvm (token)
+  (bt:with-lock-held (**persistent-qvms-lock**)
+    (remhash token **persistent-qvms**)))
+
+(defun delete-persistent-qvm (token)
+  (with-persistent-qvm (qvm metadata) token
+    (%mark-for-deletion metadata))
+  (%remove-persistent-qvm token))
+
+(defun %lookup-persistent-qvm-locked (token)
+  (gethash token **persistent-qvms**))
+
+(defun %lookup-persistent-qvm (token)
+  (bt:with-lock-held (**persistent-qvms-lock**)
+    (%lookup-persistent-qvm-locked token)))
+
+(defun %lookup-persistent-qvm-or-lose (token)
+  (or (%lookup-persistent-qvm token)
+      (error "Failed to find persistent QVM ~D" token)))
+
+(defun %marked-for-deletion-p (metadata)
+  (gethash "delete-pending" metadata))
+
+(defun %mark-for-deletion (metadata)
+  (setf (gethash "delete-pending" metadata) t))
+
+(defun canonicalize-persitent-qvm-token (token)
```

Comment:
`persitent` -> `persistent`

Reply:
Good eye, thank you!  Through the miracle of autocomplete, I never noticed this typo :)